### ISSUE TITLE
Implementing SC trend for baseflow days

### ISSUE THREAD

**Issue by lindsayplatt** [lindsayplatt/salt-modeling-data#49]
This is a little trickier now because we have subset the time series data to only include points that occur during a non-winter month and on a day where there is a baseflow value. This means that there is some variation in the sampling frequency - there are more baseflow days in June/July than there are in October/November.

To handle the "variation in sampling frequency", I am calculating the median value per month per year for the SC values that occur on baseflow days prior to running a Seasonal Kendall test. You can see this variation when I plot SC for site `01095220` below. The open circles are all the SC values falling on a non-winter, baseflow day and the closed circles are the median values per month. When I run the Seasonal Kendall with just the median values, there is less variation in the sampling frequency (some cannot be avoided, e.g. there are no April 2020 values so we do not have a median value calculated then). This process for preparing the data for Seasonal Kendall test is from [USGS Statistical Methods in Water Resources](https://doi.org/10.3133/tm4A3), Ch 12.4.1 pg. 344.

![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/f540cc60-3f4d-43c5-90bc-a53c14f7d2ee)

<details>
<summary>Code for the plot</summary>

```r
library(targets)
library(tidyverse)

ts_data_ds <- tar_read(p5_sc_baseflow) %>% filter(site_no == '01095220')
all_data <- ts_data_ds %>% 
  mutate(month = month(dateTime)) 

monthly_data <- ts_data_ds %>% 
    mutate(year = year(dateTime),
           month = month(dateTime)) %>% 
    group_by(year, month) %>% 
    summarize(SpecCond_monthly = median(SpecCond)) %>% 
    # Pretend the year-date is a day in the middle of the month 
    # so it can be plot on the same x-axis as the daily data
    mutate(dateTime = as.Date(sprintf('%s-%s-15', year, month)))

ggplot(all_data, aes(x=dateTime, color = as.character(month))) + 
  geom_point(aes(y=SpecCond), shape=1, alpha=0.5) +
  geom_point(data=monthly_data, aes(y=SpecCond_monthly), size=2) +
  geom_line(data=monthly_data, aes(y=SpecCond_monthly)) +
  theme_bw() + 
  facet_wrap(vars(month))
```

</details>

**Comment by lindsayplatt** [comment]
When I tried to run the trend test on all the sites using this new approach with calculating the median, I ended up with some errors. This was because there are some sites that don't have enough baseflow days with SpC in order to actually calculate a trend. This made me pause and need to think about minimum criteria for being able to calculate a trend.

Thus, I explored all 306 sites with the following code. I used three criteria to decide if a site should be included:

1. Per "season" (aka month in our case) a site needed to have at an average of 2 samples per year. I took this criteria from the following statement about using medians within a Seasonal Kendall test from [USGS Statistical Methods in Water Resources](https://doi.org/10.3133/tm4A3), Ch 12.4.1 pg. 344: "If the sampling frequency is reasonably consistent and there are generally two or more samples taken in each season, then the median of the available data should be used for each season of each
year".
2. Per "season", a site needed to have at least 5 different years of samples.
3. A site needed to have at least 3 seasons that passed the two criteria above.

The following PDF shows plots for all 306 sites. Sites that did not have data which passed our criteria above were labeled with `SKIP TREND`. A total of 46 sites will not be able to be used in a trend test for baseflow, so 260 sites can be used.

[all_SpC_baseflow.pdf](https://github.com/lindsayplatt/salt-modeling-data/files/14233242/all_SpC_baseflow.pdf)

Below are selected examples from the PDF.

### Sites that *did* qualify

![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/29dc1824-3c5d-442f-990d-cb8db4395edd)

This site has plenty of data to calculate a seasonal trend using all seasons.

```r
site_season_bf_qualified %>% filter(site_no == '01388000')
  site_no  month avg_samples_per_season n_years_season
  <chr>    <dbl>                  <dbl>          <int>
1 01388000     4                   5.86             14
2 01388000     5                   8                26
3 01388000     6                   9.55             29
4 01388000     7                   9.58             31
5 01388000     8                   6.22             27
6 01388000     9                   6.43             23
7 01388000    10                   4.55             11
8 01388000    11                   4.71              7
```

![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/a27cfb01-9885-46d6-9a7b-06effa41976e)

All but one of this site's seasons can be used in the seasonal Kendall trend test.

```r
site_season_bf_qualified %>% filter(site_no == '08070200')
  site_no  month avg_samples_per_season n_years_season
  <chr>    <dbl>                  <dbl>          <int>
1 08070200     4                   7.29              7
2 08070200     5                   8.44              9
3 08070200     6                   8                12
4 08070200     7                   8                10
5 08070200     8                  10.2               5
6 08070200     9                   8.6               5
7 08070200    10                   7.4               5
8 08070200    11                  12                 3
```

### Sites that *did not* qualify

![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/592170b6-2c93-48ca-a773-1f77359c5b24)

This site only had two seasons that qualified and we required 3.
```r
site_season_bf_qualified %>% filter(site_no == '09180500')
site_no  month avg_samples_per_season n_years_season
  <chr>    <dbl>                  <dbl>          <int>
1 09180500     5                    2                1
2 09180500     6                    6.3             10
3 09180500     7                   14               15
4 09180500     8                   13.2              4
```

![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/06d98114-0f5b-41a3-88e8-5aca44a15720)

This site only had one season that qualified.

```r
site_season_bf_qualified %>% filter(site_no == '05057000')
  site_no  month avg_samples_per_season n_years_season
  <chr>    <dbl>                  <dbl>          <int>
1 05057000     5                   5                 1
2 05057000     6                  11                 1
3 05057000     7                   7.5               2
4 05057000     8                   8                 3
5 05057000     9                   7.67              3
6 05057000    10                   4                 1
7 05057000    11                   2.44              9
```

![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/c2ab3660-6248-48a2-82ae-01880390838d)

This site was very close but ultimately only 2 seasons qualified.

```r
site_season_bf_qualified %>% filter(site_no == '01408000')
  site_no  month avg_samples_per_season n_years_season
  <chr>    <dbl>                  <dbl>          <int>
1 01408000     4                   7                 3
2 01408000     5                   9.33              3
3 01408000     6                  12.2               5
4 01408000     7                   8.14              7
5 01408000     8                  14.5               4
6 01408000     9                   7                 3
7 01408000    10                   7                 1
8 01408000    11                   6                 2
```

#### Code

<details>
<summary>The plot and criteria code</summary>

```r
library(targets)
library(tidyverse)

ts_data_bf <- tar_read(p5_sc_baseflow)
sites_all <- unique(ts_data_bf$site_no)
n_sites_all <- length(sites_all)

# As noted in USGS methods for seasonal kendall, it is OK to use the median to
# calculate trend if there are "generally two or more samples taken in each season"
site_season_bf_qualified <- ts_data_bf %>% 
  mutate(year = year(dateTime), 
         month = month(dateTime)) %>% 
  group_by(site_no, year, month) %>% 
  summarize(n_samples_per_yr_season = n()) %>% 
  group_by(site_no, month) %>% 
  summarize(avg_samples_per_season = mean(n_samples_per_yr_season),
            n_years_season = n()) %>% 
  # Only keep seasons that have have more than an average of two samples
  # and at least 5 different years of data
  mutate(site_season_bf_qualified = avg_samples_per_season >= 2 & n_years_season >= 5) %>% 
  # Then, only allow a site if it had at least 3 seasons that qualified
  mutate(site_bf_qualified = sum(site_season_bf_qualified) >= 3)

site_bf_qualified <- site_season_bf_qualified %>% 
  select(site_no, site_bf_qualified) %>% 
  distinct()

pdf('all_SpC_baseflow.pdf', height = 8, width = 16)
grp_size <- 25
n_grps <- ceiling(n_sites_all/grp_size)
for(grp in 1:n_grps) {
  i_start <- ((grp-1)*grp_size)+1
  i_end <- ifelse(grp == n_grps, n_sites_all, grp*grp_size)
  sites_in_grp <- sites_all[i_start:i_end]
  
  p_data <- ts_data_bf %>% 
    filter(site_no %in% sites_in_grp) %>%
    mutate(year = lubridate::year(dateTime),
           month = lubridate::month(dateTime, label=T)) %>% 
    left_join(site_bf_qualified, by = 'site_no') %>% 
    group_by(site_no) %>% 
    mutate(facet_label = sprintf('%s (%s days across %s years)', 
                                 site_no, n(),
                                 length(unique(year)))) %>% 
    ungroup() 
  
  # Add info for putting a label in the topleft of each facet
  ts_bf_sc_sites_info <- p_data %>% 
    group_by(site_no, facet_label) %>% 
    summarize(did_qualify = unique(site_bf_qualified),
              maxSpC = max(SpecCond, na.rm=TRUE), 
              minDate = min(dateTime, na.rm = TRUE),
              medianSpC = median(SpecCond, na.rm = TRUE),
              SpC95 = quantile(SpecCond, probs = 0.95, na.rm=TRUE),
              .groups = 'keep') %>% 
    mutate(site_label = ifelse(!did_qualify, 'SKIP TREND', NA)) 
  
  p <- ggplot(p_data, aes(x=dateTime, y=SpecCond)) + 
    geom_point(aes(color = month), size=0.60) +
    facet_wrap(vars(facet_label), scales='free') +
    # Add labels if it did not meet minimum criteria for calculating a trend
    geom_label(data = ts_bf_sc_sites_info,  
               aes(x = minDate, y = maxSpC, label = site_label),
               fill = 'salmon', hjust=0, nudge_y = -0.1) +
    theme_bw() + 
    xlab('Date') +
    ylab('Specific conductance, uS/cm@25degC') +
    ggtitle('SpC on baseflow days')
  
  print(p)
}
dev.off()
```

</details>

**Comment by lindsayplatt** [comment]
Using `Kendall::SeasonalMannKendall()` and mean monthly SC for any site that qualified, here are the results: 

[all_SpC_baseflow_trends_mean_pkgKendall.pdf](https://github.com/lindsayplatt/salt-modeling-data/files/14233643/all_SpC_baseflow_trends_mean_pkgKendall.pdf)

```
tar_read(p5_sc_baseflow_trend) %>% 
  group_by(attr_Trend) %>% 
  tally()

  attr_Trend     n
1 negative      24
2 none         103
3 positive     133
```

**Comment by lindsayplatt** [comment]
Same as above except filtering out data that did not qualify for a given site AND season combination. Based on counts, the only difference from the code above was that two sites shifted from no trend to a positive trend, but actually there was a lot more movement than that (see bullets below).

```
  attr_Trend     n
1 negative      24
2 none         101
3 positive     135
```

Sites that flipped from no trend to negative: `03058000`
Sites that flipped from a negative trend to no trend: `04208000`
Sites that flipped from a positive trend to no trend: `01571820`, `05055300`, `05055400`, `05059300`, `09149500`
Sites that flipped from no trend to a positive trend: `01104370`, `01518700`, `01645762`, `01673000`, `02101726`, `02336000`, `08052700`

**Comment by lindsayplatt** [comment]
Adjusting the trend method slightly by using `median` monthly values in the trend calc instead of `mean` based on methods information from [USGS Statistical Methods in Water Resources](https://doi.org/10.3133/tm4A3), Ch 12.4.1 pg. 344. Once again, the total counts changed but there were some shifts within these that aren't just one directional.

[all_SpC_baseflow_trends_median_pkgKendall.pdf](https://github.com/lindsayplatt/salt-modeling-data/files/14233713/all_SpC_baseflow_trends_median_pkgKendall.pdf)

```r
  attr_Trend     n
1 negative      23
2 none         105
3 positive     132
```

- Sites that flipped from a negative trend to no trend: `02336360`
- Sites that flipped from a positive trend to no trend: `01673000`, `02101726`, `02337170`, `03075070`, `14179000`
- Sites that flipped from no trend to a positive trend: `01645704` and `02204037`

**Comment by lindsayplatt** [comment]
Switch methods to use seasonal kendall test within `EnvStats` as recommended by [USGS Statistical Methods in Water Resources](https://doi.org/10.3133/tm4A3), Ch 12.4.1 pg. 344. The differences were quite drastic - I am not going to list all the ones that are different. Instead, you could open the two PDFs and flip back and forth to compare.

[all_SpC_baseflow_trends_median_pkgEnvStats.pdf](https://github.com/lindsayplatt/salt-modeling-data/files/14233773/all_SpC_baseflow_trends_median_pkgEnvStats.pdf)

```r
  attr_Trend     n
1 negative      62
2 none         123
3 positive      75
```

**Comment by lindsayplatt** [comment]
I am trusting the `EnvStats` package more given the source of the recommendation. Here are some examples of changes between the two packages (both using medians):

| Pkg `Kendall` | Pkg `EnvStats` |
| -------- | ------- |
| ![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/93d18780-aab8-4ea6-9043-ee52196e59ba) | ![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/2995d966-48dc-4f5b-9331-1a22e70a8b51)  |
|  
![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/b90a7f49-2eaf-4ed6-8631-642c5c259b8d)
 | 
![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/aafcfea9-5a09-417a-a48c-88bcadffd80e) |
![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/ffe7168a-f5b2-41e1-8b8a-496ba9afa3c9) | 
![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/fb657223-d633-4767-be9b-40a7c213fc48) |

**Comment by lindsayplatt** [comment]
Now a bit of a dive into SpC time series per month for each of the trends.

### Sites that have positive trends

![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/90513c3a-4ab1-48bf-a4c6-f6acf50d07f9)

![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/69bc0d13-0345-49f6-b58a-a3843ac4a763)

![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/dece0184-1986-4bbd-8725-faca104c199c)

### Sites that have no trends

![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/e5a6f294-8972-4a5b-a48f-31a56c45e7b6)

![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/a8a31eda-0d1d-4b52-b9dd-b2621a1d14f2)

![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/f7713f32-8ff4-425f-972c-8215729e1142)

### Sites that have negative trends

![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/b32a3bec-204c-40b7-b9a4-1202ebb1b85a)

![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/2f995913-504d-4e27-b411-9b7ef33e6560)

![image](https://github.com/lindsayplatt/salt-modeling-data/assets/13220910/86beaa39-a92f-413b-8d20-5d06a537f3b4)

<details>
<summary>Code used to make each plot by changing the `site_to_plot` value</summary>

```r
library(targets)
library(tidyverse)

site_to_plot <- '01608500'

ts_data_ds <- tar_read(p5_sc_baseflow_qualified) %>% 
  filter(site_no == site_to_plot) %>% 
  mutate(month = month(dateTime)) %>% 
  left_join(tar_read(p5_sc_baseflow_trend), by = 'site_no')

monthly_data <- ts_data_ds %>% 
  mutate(year = year(dateTime),
         month = month(dateTime)) %>% 
  group_by(year, month) %>% 
  summarize(SpecCond_monthly = median(SpecCond)) %>% 
  # Pretend the year-date is a day in the middle of the month 
  # so it can be plot on the same x-axis as the daily data
  mutate(dateTime = as.Date(sprintf('%s-%s-15', year, month)))

ggplot(ts_data_ds, aes(x=dateTime, color = as.character(month))) + 
  geom_point(aes(y=SpecCond), shape=1, alpha=0.5) +
  geom_point(data=monthly_data, aes(y=SpecCond_monthly), size=2) +
  geom_line(data=monthly_data, aes(y=SpecCond_monthly)) +
  theme_bw() + 
  facet_wrap(vars(month)) + 
  theme(legend.position = "none") + 
  ggtitle('Baseflow SpC for qualified months',
          subtitle = sprintf('%s, trend = %s', 
                             unique(ts_data_ds$site_no),
                             unique(ts_data_ds$attr_Trend)))
```

</details>

**Comment by lindsayplatt** [comment]
Closed by #50
